### DISCUSSION TITLE
hurfbuzz

### DISCUSSION THREAD

**Discussion by Arvant** [axmolengine/axmol#1185] [General]
now when we use Arabic character in axmol it not show correct. for language like Arabic other projects use [Harfbazz  ](https://github.com/harfbuzz/harfbuzz)lib that solve this problem and used in large projects like Android, firefox, chrome, etc...
is it possible to add this library to axmol too?

**Comment by rh101** [comment]
> now when we use Arabic character in axmol it not show correct. 

What do you mean by "now when we use..."? Are you implying that it was rendering correctly at some point, and now it is no longer working? Or, did you mean to state that Cocos2d-x/Axmol have never rendered Arabic characters in the way that you expect it to, and you are asking if it would be possible to add that functionality?

**Reply by Arvant** [reply]
Arabic just works fine on system font createWithSystemFont (Cocos2d-x/Axmol).  it would be possible to add that functionality?

**Reply by DelinWorks** [reply]
createWithSystemFont works because the system (windows or android or ios) does arabic font shaping and gives you the text back.

I've taken the time to make a single header library that renders arabic text correctly in axmol, maybe this will fullfill your needs https://github.com/DelinWorks/arabic-text-renderer-cpp-axmol it works really will and it supports vowels too!

**Reply by Arvant** [reply]
Thank you Delin. i test your project but it have problem with other Arabic like languages( Persian, Kurdish,...) and it need to modify for new unicodes.  I think that use harfbuzz is the better solution.

**Reply by DelinWorks** [reply]
if you know the letters of persian and kurdish and their connectives you can contribute and add them to the list and create a function to select which language, but yeah harfbuzz is better if you know how to integrate it in the engine!

**Reply by DelinWorks** [reply]
@Arvant this page is helpful https://harfbuzz.github.io/a-simple-shaping-example.html, we can add HarfBuzz as an optional extension into the engine and create a simple abstract shaper implementation that goes hand-in-hand with the label class, we can then add an option into a label object to enable or disable shaping to gain performance if arabic like languages are not used!

**Reply by DelinWorks** [reply]
There are of course bumps & problems that can be solved, HarfBuzz works only by providing complete arabic text, you can't have english text with arabic text or else it will break, or so I think, It needs testing though, for this case we will need to use fribidi to allow bidirectional text! or this can all be avoided by using axmol's RichText class, but that's not good for performance as it will increase draw calls, you should read this https://harfbuzz.github.io/what-harfbuzz-doesnt-do.html
